File: server/Connector.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Text;

namespace RefBox
{
	/// <summary>
	/// Represents the method that will handle the MessageReceived event of a connector
	/// </summary>
	/// <param name="connector">The Connector object that rises the event</param>
	/// <param name="source">The IPEndPoint where message comes from</param>
	/// <param name="message">The message received</param>
	public delegate void MessageReceivedEventHandler(Connector connector, IPEndPoint source, string message);

	/// <summary>
	/// Implements UDP connectivity
	/// </summary>
	public class Connector
	{
		#region Variables

		/// <summary>
		/// The input port.
		/// </summary>
		public const int portIn = 3000;
		/// <summary>
		/// The output port.
		/// </summary>
		public const int portOut = 3001;

		/// <summary>
		/// Udp listener on input port
		/// </summary>
		private UdpClient listener;

		#endregion

		#region Constructor

		/// <summary>
		/// Initializes a new instance of the <see cref="RefBox.Connector"/> class.
		/// </summary>
		public Connector ()
		{

		}

		#endregion

		#region Properties
		#endregion

		#region Events

		/// <summary>
		/// Occurs when a message is received.
		/// </summary>
		public event MessageReceivedEventHandler MessageReceived;

		#endregion

		#region Methods

		/// <summary>
		/// Broadcast the specified signal.
		/// </summary>
		/// <param name="signal">The signal to send.</param>
		public void Broadcast(Signal signal){
			if (listener == null)
				return;

			// Serialize signal
			string serialized = Serializer.Serialize (signal);
			// Get DGRAM
			byte[] dgram = ASCIIEncoding.UTF8.GetBytes (serialized + "\x03");
			// Get all ip addresses
			IPAddress[] addresses = Dns.GetHostAddresses (Dns.GetHostName ());
			// Over each IP addresses
			for(int i = 0; i < addresses.Length; ++i)
			{
				byte[] bAddress = addresses [i].GetAddressBytes ();
				bAddress [3] = 255;
				IPAddress address = new IPAddress (bAddress);
				IPEndPoint ep = new IPEndPoint (address, portOut);
				listener.Send (dgram, dgram.Length, ep);
			}
		}

		/// <summary>
		/// Raises the message received event.
		/// </summary>
		/// <param name="message">The message's source.</param>
		/// <param name="message">The message.</param>
		protected void OnMessageReceived(string message, IPEndPoint source){
			try{
				if(this.MessageReceived != null)
					MessageReceived(this, source, message);
			}catch{}
		}

		private void ReceiveCallback(IAsyncResult result)
		{
			if (listener == null)
				return;
			IPEndPoint clientEP = new IPEndPoint (IPAddress.Any, portIn);
			byte[] data = this.listener.EndReceive (result, ref clientEP);
			this.listener.BeginReceive (ReceiveCallback, null);
			string sData = ASCIIEncoding.UTF8.GetString (data);
			OnMessageReceived(sData, clientEP);
		}

		/// <summary>
		/// Starts async reception of UDP messages
		/// </summary>
		public void Start(){
			if(listener != null)
				return;
			this.listener = new UdpClient (portIn);
			this.listener.EnableBroadcast = true;
			this.listener.DontFragment = true;
			this.listener.BeginReceive(ReceiveCallback, null);
		}

		/// <summary>
		/// Stops async reception of UDP messages
		/// </summary>
		public void Stop(){
			this.listener.Close ();
			this.listener = null;
		}

		#endregion
	}
}

File: refbox_rosclient/src/Connector.cpp
#include "Connector.h"
#include "Serializer.h"
using namespace refbox;
using boost::asio::ip::udp;

// const int Connector::inputBufferSize(81920);
// const int Connector::portIn(3001);
// const int Connector::portOut(3000);

Connector::Connector ()	: running_(false), mainThread_(NULL),
	socket_(NULL), inputBuffer_(NULL){
	this->inputBuffer_ = new char[inputBufferSize];
}

void Connector::beginReceive(){
	this->socket_->async_receive_from(
		boost::asio::buffer(inputBuffer_, inputBufferSize),
		this->senderEP_,
		boost::bind(
			&Connector::receiveCallback,
			this,
			boost::asio::placeholders::error,
			boost::asio::placeholders::bytes_transferred
		)
	);
}

void Connector::mainThreadTask(){
	while( this->running_ ){
		this->service_.run();
	}
}

void Connector::onMessageReceived(const std::string& message, const EndPoint& source){
	try{
		this->messageReceived(this, source, message);
	}catch(...){
#if DEBUG
		std::cout<< "Error!" << std::endl;
#endif
	}
}

void Connector::receiveCallback(const boost::system::error_code& error, size_t received){
	if(error || (received < 1))
		return;
	EndPoint clientEP(this->senderEP_);
	std::string sData(this->inputBuffer_, received);
	this->beginReceive();
	this->onMessageReceived(sData, clientEP);
}

void Connector::start(){
	if(this->running_)
		return;
	this->socket_ = new udp::socket(this->service_, EndPoint(udp::v4(), portIn));
	this->socket_->set_option(udp::socket::reuse_address(true));
	this->socket_->set_option(boost::asio::socket_base::broadcast(true));
	this->beginReceive();
	this->running_ = true;
	this->mainThread_ = new boost::thread(boost::bind(&Connector::mainThreadTask, this));
}

void Connector::stop(){
	if(!this->running_)
		return;
	this->running_ = false;
	this->mainThread_->join();
	// this->service.reset();
	this->service_.stop();
	delete this->socket_;
	this->socket_ = NULL;
	delete this->mainThread_;
	this->mainThread_ = NULL;
}

void Connector::broadcast(const Event& event) const{
	if(!this->running_ || (this->socket_ == NULL))
		return;
	// Serialize signal
	std::string serialized = Serializer::serialize (event);
	// Send the data
	this->socket_->send_to(
		boost::asio::buffer(serialized.c_str(), serialized.size()),
		EndPoint(udp::v4(), portOut)
	);
}